client/core/core.go

```diff
@@ -608,6 +608,19 @@ func (dc *dexConnection) reconcileTrades(srvOrderStatuses []*msgjson.OrderStatus
 		previousStatus := trade.metaData.Status
 		newStatus := order.OrderStatus(srvOrderStatus.Status)
 		trade.metaData.Status = newStatus
+		// If there is an associated cancel order, and we are revising the
+		// status of the targeted order, we can infer the cancel order is done.
```

Comment:
Only orders that were previously active (Epoch / Booked) can get here, either from a direct reconciliation using the status returned from the `connect` response or using the `order_status` route. Regardless of how this update is triggered, **stale** cancel orders for such knownActive trades are deleted, before the update takes place.

It'd be more accurate to note here _(but don't know if it's really important or better to leave it to reader's inference)_ that any associated cancel order would have been deleted if it had stayed at Epoch status longer than 2 epochs, and this specifically targets cancel orders whose epoch may not have passed.

Reply:
That's not it though.  I got here testing another PR that made the server revoke the epoch status trade without telling client, so that the associated cancel would never have a reference to it.  That's also where the other block in https://github.com/decred/dcrdex/pull/786/files#diff-13a4a6b12c1ae844ef90151638f194427cfdc545fceb58d52c0f6f33b81d859bR3121 helps.

Reply:
Noting these lines: https://github.com/decred/dcrdex/pull/786/files#diff-13a4a6b12c1ae844ef90151638f194427cfdc545fceb58d52c0f6f33b81d859bR650-R657
It couldn't even get to `deleteStaleCancelOrder` in the case I was testing because the trade wasn't in `srvActiveOrderStatuses`.

Reply:
Yeah, retiring obsolete cancels on load from db is vital, some cancels will just not get retired naturally, as you've seen and pointed out.

I'm not able to visit the code section referenced in the link in your second comment, but looking at the code again, you're right. Trades not in `srvActiveOrderStatuses` don't get to `deleteStaleCancelOrder`.